File: src/main/java/com/timeback/ui/feature/report/ReportViewModel.java
package com.timeback.ui.feature.report;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.timeback.ui.domain.gateway.FeatureGateway;
import com.timeback.ui.domain.model.ReportPeriod;
import com.timeback.ui.domain.model.ScreenState;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

/**
 * UI-07 리포트 ViewModel.
 * 일간/주간/월간 리포트를 조회한다.
 */
@HiltViewModel
public class ReportViewModel extends ViewModel {

    private final FeatureGateway gateway;
    private final MutableLiveData<ScreenState> _state = new MutableLiveData<>(new ScreenState.Initial());
    public final LiveData<ScreenState> state = _state;

    private ReportPeriod currentPeriod = ReportPeriod.WEEKLY;

    @Inject
    public ReportViewModel(FeatureGateway gateway) {
        this.gateway = gateway;
        loadReport(currentPeriod);
    }

    public void loadReport(ReportPeriod period) {
        this.currentPeriod = period;
        _state.setValue(new ScreenState.Loading());
        gateway.readReport(period).thenAccept(_state::postValue);
    }

    public void changePeriod(ReportPeriod period) {
        loadReport(period);
    }
}

File: src/main/java/com/timeback/ui/feature/report/ReportViewModel.kt
package com.timeback.ui.feature.report

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.timeback.ui.domain.gateway.FeatureGateway
import com.timeback.ui.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI-07 리포트 ViewModel
 * CT-04 §10: ReadReport, ChangeReportPeriod
 */
@HiltViewModel
class ReportViewModel @Inject constructor(
    private val gateway: FeatureGateway
) : ViewModel() {

    private val _state = MutableStateFlow<ScreenState>(ScreenState.Initial)
    val state: StateFlow<ScreenState> = _state.asStateFlow()

    private var currentPeriod = ReportPeriod.WEEKLY

    init {
        loadReport(currentPeriod)
    }

    fun loadReport(period: ReportPeriod) {
        currentPeriod = period
        viewModelScope.launch {
            _state.value = ScreenState.Loading
            _state.value = gateway.readReport(period)
        }
    }

    fun changePeriod(period: ReportPeriod) {
        if (period != currentPeriod) {
            loadReport(period)
        }
    }
}
